specs/electra/beacon-chain.md

```diff
@@ -631,12 +634,19 @@ def switch_to_compounding_validator(state: BeaconState, index: ValidatorIndex) -
 
 ```python
 def queue_excess_active_balance(state: BeaconState, index: ValidatorIndex) -> None:
+    validator = state.validators[index]
     balance = state.balances[index]
     if balance > MIN_ACTIVATION_BALANCE:
         excess_balance = balance - MIN_ACTIVATION_BALANCE
         state.balances[index] = MIN_ACTIVATION_BALANCE
-        state.pending_balance_deposits.append(
-            PendingBalanceDeposit(index=index, amount=excess_balance)
+        state.pending_deposits.append(
+            PendingDeposit(
+                epoch=get_current_epoch(state),
```

Comment:
Is there an off-by-one here?

When epoch N finalizes, everything up through and including `compute_start_slot_at_epoch(N)` is finalized.

1. If activations are processed for `deposit.epoch <= finalized_checkpoint.epoch`, then that would also activate deposits on the non-finalized slots within N (only the initial slot of N is finalized).
2. If activations are processed for `deposit.epoch < finalized_checkpoint.epoch`, then deposits within the very first slot of N are not processed for another ~6 minutes despite already being finalized.

Currently, it seems we are following the second strategy (`deposit.epoch >= state.finalized_checkpoint.epoch`). Probably fine to have the extra ~6 minutes wait for deposits within the first slot, but could be avoided, so still keeping the comment for awareness.

Reply:
I personally in favour of strategy (2) because of its simplicity

Reply:
I thought about this and I changed my mind. The issue is that clients may (and probably will) filter deposits by blocks or db invariants like "finalized", the presence of this off by one creates edge cases that need to be always checked and tested. While it's not likely to create a consensus bug I feel it will be a source of bugs nonetheless

Reply:
If we set `deposit.epoch` in the queue to `get_current_epoch(state) + 1`, will that prevent issues that you have in mind?

Reply:
I guess storing `slot` and processing deposits up through `deposit_receipt.slot <= compute_start_slot_at_epoch(state.finalized_checkpoint.epoch)` is the most straight forward.